rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,504 @@
+  | Scheduled (* Is in the `transaction_pool` and not in the `transition_frontier`. *)
+  | Unknown
+```
+
+We also have the `receipt_chain_hash` object and is used to prove a `user_command` made it into a block. It does this by forming a Merkle list of `user_command`s from some `proving_receipt` to a `resulting_receipt`. The `resulting_receipt` is typically the `receipt_chain_hash` of a user's account on the best tip ledger. More info about `receipt_chain_hash` can be found on this [RFC](rfcs/0006-receipt-chain-proving.md) and this [OCaml file](../src/lib/receipt_chain_database_lib/intf.ml).
+
+Below is an OCaml API of involving `receipt_chain_hash` type:
+
+```ocaml
+module Receipt_chain_hash = struct
+  type t = {
+    hash: string
+    previous_hash: string
+  }
+
+  val compute_receipt : ~prev:t -> User_command.t -> t
+end
+```
+
+These containers should also have performant pagination queries, which is essentially a single contiguous range of sorted elements. `external_transitions` and `transactions` should be ordered by block length. Previously, we sorted these objects by time. Block length can arguably be a better ordering for transitions because clients can receive transitions at different times, but what remains constant is the block length that the transition appears in the blockchain. Additionally, we can sort transactions by the minimum block length of the `external_transition`s that they appear in. Typically, a client would expect to receive a transaction at the soonest block, which is the block with the smallest block length. If they do not appear in an `external_transition`, they will not appear in the sorting. A drawback with sorting `external_transitions` and transaction based on date is that if the client wants to add an arbitrary transaction and `external_transition` and persist them, then they would not have the extra burden of tagging these objects with a timestamp for pagination. Block length alleviates this use.
+
+## Implementation
+
+This section will discuss a proposal to make the querying done on a different process, which will be called the client process, and it will take advantage of in-memory data structures as well as persistent databases.
+
+A persistent database that we would like to introduce into our codebase is `SQLite`. `SQLite` would be the best relational database for our use case because it has an Apache 2 license and its a very small binary (3.7MB on my Macbook). `SQLite` has a considerably good amount of support for OCaml compared to the other databases that will be discussed and it would be relatively easy to get `SQLite` working off. There is even a nice OCaml library called [ocaml-sqlexpr](https://github.com/mfp/ocaml-sqlexpr) that enables us to embed type-safe `SQLite` queries onto our OCaml code.
```

Comment:
I think the original idea around SQLite was thinking that we would have this in process or at least part of the distributable we ship of coda.

Given that this will live out of process and out of that distributable, you’re right that we should consider other databases